File: python/proverb.py
def proverb(*lst, qualifier=None):
    if lst == ():
        return []

    res = []

    for i in range(len(lst) - 1):
        sublst = lst[i:i + 2]
        res.append(f'For want of a {sublst[0]} the {sublst[1]} was lost.')

    if qualifier is None:
        res.append(f'And all for the want of a {lst[0]}.')
    else:
        res.append(f'And all for the want of a {qualifier} {lst[0]}.')
    return res

File: zig/proverb.zig
const std = @import("std");
const mem = std.mem;
const fmt = std.fmt;

pub fn recite(allocator: mem.Allocator, words: []const []const u8) (fmt.AllocPrintError || mem.Allocator.Error)![][]u8 {
    var lst = try allocator.alloc([]u8, words.len);
    if (words.len == 0) {
        return lst;
    }

    for (1..words.len) |i| {
        lst[i - 1] = try fmt.allocPrint(
            allocator,
            "For want of a {s} the {s} was lost.\n",
            .{ words[i - 1], words[i] },
        );
    }

    lst[lst.len - 1] = try fmt.allocPrint(
        allocator,
        "And all for the want of a {s}.\n",
        .{words[0]},
    );
    return lst;
}
